wit/request-reply.wit

```diff
@@ -0,0 +1,27 @@
+/// The request-reply interface allows a guest to send a message and await a response. This
+/// interface is considered optional as not all message services support the concept of
+/// request/reply. However, request/reply is a very common pattern in messaging and as such, we have
+/// included it as a core interface.
+interface request-reply {
+    use types.{client, message, error};
+
+    /// Options for a request/reply operation
+    record request-options {
+        /// The maximum amount of time to wait for a response. If the timeout value is not set, then
+        /// the request/reply operation will block until a message is received in response.
+        timeout-ms: option<u32>,
+        /// The number of expected replies. If the number of replies is not set, then the
+        /// request/reply operation will block until a single message is received in response (or a
+        /// timeout value is reached). This can be used in conjunction for "scatter/gather" type
+        /// operations.
+        expected-replies: option<u32>,
+    }
+
+    /// Performs a blocking request/reply operation with an optional set of request options
+    request: func(c: client, msg: message, opts: option<request-options>) -> result<option<list<message>>, error>;
+
+    /// Replies to the given message with the given response message. The details of which channel 
+    /// the message is sent to is up to the implementation. This allows for reply to details to be
+    /// handled in the best way possible for the underlying messaging system.
+    reply: func(reply-to: borrow<message>, reply: message) -> result<_, error>;
```

Comment:
What if instead of `reply` being an imported interface, it was instead an exported interface:
```wit
interface incoming-request-reply-handler {
  handle: func(msg: message) -> result<message, error>;
}
```
Here, the reply is just the success case of the `result`.  This can avoid some of the weird spec interaction questions like I was pointing out in my other comment on `complete`.  Thus, there could be a request-reply world that looks like:
```wit
world request-reply-messaging {
  include messaging;
  import outgoing-request-reply-handler;
  export incoming-request-reply-handler;
}
```

Reply:
So I actually think this would make things too complicated. This would make it so a component would have implement a different `handle` function if it is responding to requests. The `reply` function here is for a handler to use to reply to a message it has received, no matter the source. The host could throw it away or decide not to send it (for example, in the code I've written for NATS, if reply wasn't set, we just don't send anything back). That could still work here, but I don't like the idea of having another interface the guest has to export rather than just sending the reply and letting the implementation decide what to do with it (which feels more compatible)

Reply:
Maybe you could help me understand how these request-reply messaging APIs are used in practice because it feels like there's something I'm missing.  I guess a basic question is: is calling the `reply` method on a `message` passed as the parameter to `handle` able to express some use cases that returning the reply as a return value cannot?  If so: what are they?  If not, how is ye olde parameters+results not the simpler interface?

If there is a good reason for having `reply` be a function, I think the spec text needs to go into more detail about what happens in various cases such as:
* What is the interaction between the return value of `handle` (the `result`) and `reply`: if I `reply` but then return an `error` (signalling that this message should not be considered "procssed", per the current spec text), does that mean `handle` might get called again for the same `message` and, in that case, will the host care of making sure there is only one `reply` or does that guest have to worry about that?
* If I destroy a `message` without calling `reply`, is some default reply sent (that would be received by a blocking `request` caller) and if so, what is it?
* What happens if I call `reply` multiple times for the same `message`?
* If I `reply` to a `message` that was sent by `send` (not `request`) what happens?  Is there any way to detect whether a `message` is "replyable"?
* If I forward a `message` (to any of `send`, `request` or `reply`) and *that* recipient calls `reply`, who does the reply go to?

I think all these questions go away if replies are return values, because there are just less moving parts (which suggests that it is indeed the simpler interface).

Reply:
FWIW, I originally had the idea of an optional `reply-to` field than then the handler can just call the `producer.send` function to send a message back, but got push back on it initially. To me there seems to be 3 options:

1. Have this be in the `metadata` of the message and entirely external to the interface
2. Optional reply to field that a handler can use with `producer.send`
3. A reply-request interface

My opinions on the 3:

1. Bad idea because every implementor may or may not have it and the naming is not standardized
2. Nicer because it is named, but if you want to depend on it, the implementation might not provide it
3. Best because an implementation can indicate if it supports replying. This is why the details of how to reply are left up to said implementation rather than being a field on the message